### ISSUE TITLE
how to decide which glyphs should be considered "mark", "base" or "ligature"?

### ISSUE THREAD

**Issue by anthrotype** [googlei18n/ufo2ft#261]
This issue https://github.com/googlei18n/noto-fonts/issues/1223 prompted me to reconsider the way the MarkFeatureWriter currently classifies glyphs into marks, bases or ligatures.

The way Glyphs.app does it, is by using its GlyphData.xml database, where hardcoded glyph names map to some Unicode properties and a predefined list of anchor names that each can have. Glyph names that are not contained in that global database can be classified via the glyph property panel (CMD+ALT+I) which stores these override values inside the source file.
In glyphsLib we use this information to generate a GDEF GlyphClassDef table in the feature file.

However, in the ufo2ft MarkFeatureWriter, the way we classify glyphs into these three categories is simply the following:
- we check for the presence of an underscore-prefixed anchor, e.g. "_top", and if that has some corresponding anchor named "top" anywhere else in the font, we automatically consider that glyph a combining mark.
- anything that does not have an underscore prefixed anchor, but has some non-underscore-prefixed anchors, e.g. "top", we consider a base glyph
- anything that has a base anchor with some numbered suffix, e.g. "top_1", "top_2", we consider a ligature glyph and include in the mark2liga lookup.

We don't take into account the Unicode properties of the glyph involved, nor we read the GDEF GlyphClassDef which the user (or glyphsLib) may have defined in the features.fea.

This leads to, e.g., spacing diacritic glyphs -- that by definition are supposed to _not_ be considered combining marks --, be erroneously treated as mark glyphs in the `mark` feature, just because they happen to have a `_top` anchor. But the designer may have placed an anchor in there because she wanted to reuse that glyph as a component in an accented composite glyph.

When exporting from, say, Glyphs.app, those glyphs won't be considered "mark" glyphs because their "subCategory" in the GlyphData.xml is "Spacing".

We have to think of a way to determine/store these glyph categories that we could generalise to both the Glyphs and the UFO workflow, which is not tied too much with the former.

Maybe the MarkFeatureWriter should parse the GDEF GlyphClasDef statement in the feature file and use that to group the glyphs into these three non-overlapping sets.

comments?

**Comment by m4rc1e** [comment]
I don't know if this will cover all edge cases but couldn't we use the unicode category for each glyph?

```
>>> import unicodedata
>>> unicodedata.category(u'\u00B4') # ACUTE ACCENT
Sk
>>> unicodedata.category(u'\u0300')  # COMBINING GRAVE ACCENT
Mn 
>>> unicodedata.category(u'\u0944')  # Devanagari Vowel Sign Vocalic Rr
Mn
>>> unicodedata.category(u'\u0650') # ARABIC KASRA
Mn
```

If the category is `Mn` it's a mark?

**Comment by anthrotype** [comment]
I would prefer not to read the unicode category (think about those glyphs that do not have a unicode codepoint assigned).

Maybe we could this:
1) if the features.fea contains a GDEF GlyphClassDef statement, we use that to categorize the glyphs into combining marks, base and ligature glyphs. We consider accent anchors not any "_" prefixed anchor found anywhere, but any "_"-prefixed anchor found inside the glyphs that are defined as "combining marks" given the GDEF GlyphClassDef. GlyphsLib will generate a GDEF from the GlyphData.xml (or per-glyph property overrides) when creating UFOs from .glyphs source file.
2) if features.fea does not contain a GDEF GlyphClassDef statement, we keep using the current greedy algorithm.

**Comment by MrBrezina** [comment]
I like reading this information from the GDEF. It makes perfect sense. Re the second step, I am not sure what is your official stand on auto-magic solutions, but that is probably best done in a separate tool. I would not generate anything if the GDEF is not there unless it breaks too many existing workflows. Or add a flag `automatically_from_anchors` to turn that option on.

**Comment by anthrotype** [comment]
Fixed with #276
